### PR TITLE
tls: improve TLS handshake error message

### PR DESCRIPTION
**Pull request by pwhelan**
Errors during the TLS handshake are not reported or are obscure and hard to understand. Now we report the error:

  [2022/01/03 14:21:24] [error] [tls] error: unexpected EOF

The error handling code for the unexpected EOF error has to be written in this way due to a legacy bug in OpenSSL 1.x, as mentioned at https://www.openssl.org/docs/man1.1.1/man3/SSL_get_error.html.

----

**Testing**
Before we can approve your change; please submit the following in a comment:
- [ ] Debug log output from testing the change

- [ ] Attached [Valgrind](https://valgrind.org/docs/manual/quick-start.html) output that shows no leaks or memory corruption was found

----

### CODE REVIEW

**Comment by pwhelan** [comment]
```
valgrind ./bin/fluent-bit -vvv -i dummy -o http -p tls=on -p host=127.0.0.1 -p port=7777 -p match='*' -f 1
==46955== Memcheck, a memory error detector
==46955== Copyright (C) 2002-2017, and GNU GPL'd, by Julian Seward et al.
==46955== Using Valgrind-3.18.1 and LibVEX; rerun with -h for copyright info
==46955== Command: ./bin/fluent-bit -vvv -i dummy -o http -p tls=on -p host=127.0.0.1 -p port=7777 -p match=* -f 1
==46955== 
Fluent Bit v1.9.0
* Copyright (C) 2019-2021 The Fluent Bit Authors
* Copyright (C) 2015-2018 Treasure Data
* Fluent Bit is a CNCF sub-project under the umbrella of Fluentd
* https://fluentbit.io

[2022/01/03 14:21:11] [ info] Configuration:
[2022/01/03 14:21:11] [ info]  flush time     | 1.000000 seconds
[2022/01/03 14:21:11] [ info]  grace          | 5 seconds
[2022/01/03 14:21:11] [ info]  daemon         | 0
[2022/01/03 14:21:11] [ info] ___________
[2022/01/03 14:21:11] [ info]  inputs:
[2022/01/03 14:21:11] [ info]      dummy
[2022/01/03 14:21:11] [ info] ___________
[2022/01/03 14:21:11] [ info]  filters:
[2022/01/03 14:21:11] [ info] ___________
[2022/01/03 14:21:11] [ info]  outputs:
[2022/01/03 14:21:11] [ info]      http.0
[2022/01/03 14:21:11] [ info] ___________
[2022/01/03 14:21:11] [ info]  collectors:
[2022/01/03 14:21:11] [ info] [engine] started (pid=46955)
[2022/01/03 14:21:11] [debug] [engine] coroutine stack size: 24576 bytes (24.0K)
[2022/01/03 14:21:11] [debug] [storage] [cio stream] new stream registered: dummy.0
[2022/01/03 14:21:11] [ info] [storage] version=1.1.5, initializing...
[2022/01/03 14:21:11] [ info] [storage] in-memory
[2022/01/03 14:21:11] [ info] [storage] normal synchronization mode, checksum disabled, max_chunks_up=128
[2022/01/03 14:21:11] [ info] [cmetrics] version=0.2.2
[2022/01/03 14:21:11] [debug] [http:http.0] created event channels: read=19 write=20
[2022/01/03 14:21:11] [debug] [router] match rule dummy.0:http.0
[2022/01/03 14:21:11] [ info] [sp] stream processor started
[2022/01/03 14:21:12] [debug] [input chunk] update output instances with new chunk size diff=26
[2022/01/03 14:21:13] [debug] [task] created task=0x53e6b00 id=0 OK
[2022/01/03 14:21:13] [debug] [input chunk] update output instances with new chunk size diff=26
[2022/01/03 14:21:14] [debug] [task] created task=0x5469400 id=1 OK
[2022/01/03 14:21:14] [debug] [input chunk] update output instances with new chunk size diff=26
[2022/01/03 14:21:15] [debug] [task] created task=0x54e4100 id=2 OK
[2022/01/03 14:21:15] [debug] [input chunk] update output instances with new chunk size diff=26
[2022/01/03 14:21:16] [debug] [task] created task=0x555ee00 id=3 OK
[2022/01/03 14:21:16] [debug] [input chunk] update output instances with new chunk size diff=26
[2022/01/03 14:21:17] [debug] [task] created task=0x55d9b00 id=4 OK
[2022/01/03 14:21:17] [debug] [input chunk] update output instances with new chunk size diff=26
[2022/01/03 14:21:18] [debug] [task] created task=0x5654800 id=5 OK
[2022/01/03 14:21:18] [debug] [input chunk] update output instances with new chunk size diff=26
[2022/01/03 14:21:19] [debug] [task] created task=0x56cf500 id=6 OK
[2022/01/03 14:21:19] [debug] [input chunk] update output instances with new chunk size diff=26
[2022/01/03 14:21:20] [debug] [task] created task=0x7066220 id=7 OK
[2022/01/03 14:21:20] [debug] [input chunk] update output instances with new chunk size diff=26
[2022/01/03 14:21:21] [debug] [task] created task=0x70e0f20 id=8 OK
[2022/01/03 14:21:21] [debug] [input chunk] update output instances with new chunk size diff=26
[2022/01/03 14:21:22] [debug] [task] created task=0x715bc20 id=9 OK
[2022/01/03 14:21:22] [debug] [input chunk] update output instances with new chunk size diff=26
[2022/01/03 14:21:23] [debug] [task] created task=0x71d6920 id=10 OK
[2022/01/03 14:21:23] [debug] [input chunk] update output instances with new chunk size diff=26
[2022/01/03 14:21:24] [error] [upstream] connection #25 to 127.0.0.1:7777 timed out after 10 seconds
[2022/01/03 14:21:24] [error] [tls] error: unexpected EOF
[2022/01/03 14:21:24] [debug] [socket] could not validate socket status for #25 (don't worry)
[2022/01/03 14:21:24] [debug] [upstream] connection #-1 failed to 127.0.0.1:7777
[2022/01/03 14:21:24] [error] [output:http:http.0] no upstream connections available to 127.0.0.1:7777
[2022/01/03 14:21:24] [error] [upstream] connection #26 to 127.0.0.1:7777 timed out after 10 seconds
[2022/01/03 14:21:24] [error] [tls] error: unexpected EOF
[2022/01/03 14:21:24] [debug] [socket] could not validate socket status for #26 (don't worry)
[2022/01/03 14:21:24] [debug] [upstream] connection #-1 failed to 127.0.0.1:7777
[2022/01/03 14:21:24] [error] [output:http:http.0] no upstream connections available to 127.0.0.1:7777
[2022/01/03 14:21:24] [debug] [task] created task=0x72516c0 id=11 OK
[2022/01/03 14:21:24] [debug] [input chunk] update output instances with new chunk size diff=26
[2022/01/03 14:21:24] [debug] [out flush] cb_destroy coro_id=0
[2022/01/03 14:21:24] [debug] [out flush] cb_destroy coro_id=1
[2022/01/03 14:21:24] [debug] [retry] new retry created for task_id=0 attempts=1
[2022/01/03 14:21:24] [ warn] [engine] failed to flush chunk '46955-1641230472.243811716.flb', retry in 10 seconds: task_id=0, input=dummy.0 > output=http.0 (out_id=0)
[2022/01/03 14:21:24] [debug] [retry] new retry created for task_id=1 attempts=1
[2022/01/03 14:21:24] [ warn] [engine] failed to flush chunk '46955-1641230473.324440044.flb', retry in 10 seconds: task_id=1, input=dummy.0 > output=http.0 (out_id=0)
^C[2022/01/03 14:21:24] [engine] caught signal (SIGINT)
[2022/01/03 14:21:24] [debug] [task] created task=0x72cc9d0 id=12 OK
[2022/01/03 14:21:25] [ warn] [engine] service will shutdown in max 5 seconds
[2022/01/03 14:21:25] [debug] [input chunk] dummy.0 is paused, cannot append records
[2022/01/03 14:21:25] [ info] [task] dummy/dummy.0 has 13 pending task(s):
[2022/01/03 14:21:25] [ info] [task]   task_id=0 still running on route(s): http/http.0 
[2022/01/03 14:21:25] [ info] [task]   task_id=1 still running on route(s): http/http.0 
[2022/01/03 14:21:25] [ info] [task]   task_id=2 still running on route(s): http/http.0 
[2022/01/03 14:21:25] [ info] [task]   task_id=3 still running on route(s): http/http.0 
[2022/01/03 14:21:25] [ info] [task]   task_id=4 still running on route(s): http/http.0 
[2022/01/03 14:21:25] [ info] [task]   task_id=5 still running on route(s): http/http.0 
[2022/01/03 14:21:25] [ info] [task]   task_id=6 still running on route(s): http/http.0 
[2022/01/03 14:21:25] [ info] [task]   task_id=7 still running on route(s): http/http.0 
[2022/01/03 14:21:25] [ info] [task]   task_id=8 still running on route(s): http/http.0 
[2022/01/03 14:21:25] [ info] [task]   task_id=9 still running on route(s): http/http.0 
[2022/01/03 14:21:25] [ info] [task]   task_id=10 still running on route(s): http/http.0 
[2022/01/03 14:21:25] [ info] [task]   task_id=11 still running on route(s): http/http.0 
[2022/01/03 14:21:25] [ info] [task]   task_id=12 still running on route(s): http/http.0 
[2022/01/03 14:21:25] [error] [tls] error: unexpected EOF
[2022/01/03 14:21:25] [debug] [socket] could not validate socket status for #27 (don't worry)
[2022/01/03 14:21:25] [debug] [upstream] connection #-1 failed to 127.0.0.1:7777
[2022/01/03 14:21:25] [error] [output:http:http.0] no upstream connections available to 127.0.0.1:7777
[2022/01/03 14:21:25] [debug] [out flush] cb_destroy coro_id=2
[2022/01/03 14:21:26] [debug] [input chunk] dummy.0 is paused, cannot append records
[2022/01/03 14:21:27] [error] [tls] error: unexpected EOF
[2022/01/03 14:21:27] [debug] [socket] could not validate socket status for #28 (don't worry)
[2022/01/03 14:21:27] [debug] [upstream] connection #-1 failed to 127.0.0.1:7777
[2022/01/03 14:21:27] [error] [output:http:http.0] no upstream connections available to 127.0.0.1:7777
[2022/01/03 14:21:27] [error] [tls] error: unexpected EOF
[2022/01/03 14:21:27] [debug] [socket] could not validate socket status for #29 (don't worry)
[2022/01/03 14:21:27] [debug] [upstream] connection #-1 failed to 127.0.0.1:7777
[2022/01/03 14:21:27] [error] [output:http:http.0] no upstream connections available to 127.0.0.1:7777
[2022/01/03 14:21:27] [debug] [input chunk] dummy.0 is paused, cannot append records
[2022/01/03 14:21:27] [debug] [out flush] cb_destroy coro_id=3
[2022/01/03 14:21:27] [debug] [out flush] cb_destroy coro_id=4
[2022/01/03 14:21:28] [debug] [input chunk] dummy.0 is paused, cannot append records
[2022/01/03 14:21:28] [error] [tls] error: unexpected EOF
[2022/01/03 14:21:28] [debug] [socket] could not validate socket status for #30 (don't worry)
[2022/01/03 14:21:28] [debug] [upstream] connection #-1 failed to 127.0.0.1:7777
[2022/01/03 14:21:28] [error] [output:http:http.0] no upstream connections available to 127.0.0.1:7777
[2022/01/03 14:21:28] [debug] [out flush] cb_destroy coro_id=5
[2022/01/03 14:21:29] [debug] [input chunk] dummy.0 is paused, cannot append records
[2022/01/03 14:21:29] [ info] [task] dummy/dummy.0 has 13 pending task(s):
[2022/01/03 14:21:29] [ info] [task]   task_id=0 still running on route(s): http/http.0 
[2022/01/03 14:21:29] [ info] [task]   task_id=1 still running on route(s): http/http.0 
[2022/01/03 14:21:29] [ info] [task]   task_id=2 still running on route(s): http/http.0 
[2022/01/03 14:21:29] [ info] [task]   task_id=3 still running on route(s): http/http.0 
[2022/01/03 14:21:29] [ info] [task]   task_id=4 still running on route(s): http/http.0 
[2022/01/03 14:21:29] [ info] [task]   task_id=5 still running on route(s): http/http.0 
[2022/01/03 14:21:29] [ info] [task]   task_id=6 still running on route(s): http/http.0 
[2022/01/03 14:21:29] [ info] [task]   task_id=7 still running on route(s): http/http.0 
[2022/01/03 14:21:29] [ info] [task]   task_id=8 still running on route(s): http/http.0 
[2022/01/03 14:21:29] [ info] [task]   task_id=9 still running on route(s): http/http.0 
[2022/01/03 14:21:29] [ info] [task]   task_id=10 still running on route(s): http/http.0 
[2022/01/03 14:21:29] [ info] [task]   task_id=11 still running on route(s): http/http.0 
[2022/01/03 14:21:29] [ info] [task]   task_id=12 still running on route(s): http/http.0 
[2022/01/03 14:21:29] [ info] [engine] service has stopped (11 pending tasks)
[2022/01/03 14:21:29] [debug] [task] destroy task=0x53e6b00 (task_id=0)
[2022/01/03 14:21:29] [debug] [retry] task retry=0x72b83d0, invalidated from the scheduler
[2022/01/03 14:21:29] [debug] [task] destroy task=0x5469400 (task_id=1)
[2022/01/03 14:21:29] [debug] [retry] task retry=0x72cc7e0, invalidated from the scheduler
[2022/01/03 14:21:29] [debug] [task] destroy task=0x54e4100 (task_id=2)
[2022/01/03 14:21:29] [debug] [task] destroy task=0x555ee00 (task_id=3)
==46955== 
==46955== HEAP SUMMARY:
==46955==     in use at exit: 805,220 bytes in 3,981 blocks
==46955==   total heap usage: 7,423 allocs, 3,442 frees, 7,002,452 bytes allocated
==46955== 
==46955== LEAK SUMMARY:
==46955==    definitely lost: 0 bytes in 0 blocks
==46955==    indirectly lost: 0 bytes in 0 blocks
==46955==      possibly lost: 696,178 bytes in 292 blocks
==46955==    still reachable: 109,042 bytes in 3,689 blocks
==46955==         suppressed: 0 bytes in 0 blocks
==46955== Rerun with --leak-check=full to see details of leaked memory
==46955== 
==46955== For lists of detected and suppressed errors, rerun with: -s
==46955== ERROR SUMMARY: 0 errors from 0 contexts (suppressed: 0 from 0)
```

**Comment by edsiper** [comment]
@pwhelan can you pls rebase on top of Master ? there are some recent openssl changes

File: src/tls/openssl.c [hunk 1]
```diff
@@ -363,6 +366,10 @@ static int tls_net_read(struct flb_upstream_conn *u_conn,
         else if (ret == SSL_ERROR_WANT_WRITE) {
             ret = FLB_TLS_WANT_WRITE;
         }
+        else if (ret < 0) {
+            ERR_error_string(ret, errbuf);
```
Comment (nokute78) [inline src/tls/openssl.c:370]: How about using `ERR_error_string_n` ?
https://www.openssl.org/docs/man3.0/man3/ERR_error_string.html
```
Note that this function is not thread-safe and does no checks on the size of the buffer; use ERR_error_string_n() instead.
```

File: src/tls/openssl.c [hunk 2]
```diff
@@ -396,6 +404,8 @@ static int tls_net_write(struct flb_upstream_conn *u_conn,
             ret = FLB_TLS_WANT_READ;
         }
         else {
+            ERR_error_string(ret, errbuf);
```
Comment (nokute78) [inline src/tls/openssl.c:407]: Ditto.

File: src/tls/openssl.c [hunk 3]
```diff
@@ -426,6 +437,15 @@ static int tls_net_handshake(struct flb_tls *tls, void *ptr_session)
         if (ret != SSL_ERROR_WANT_READ &&
             ret != SSL_ERROR_WANT_WRITE) {
             ret = SSL_get_error(session->ssl, ret);
+            // The SSL_ERROR_SYSCALL with errno value of 0 
+            // indicates unexpected EOF from the peer.
+            // This is fixed in OpenSSL 3.0.
+            if (ret == 0) {
+            	flb_error("[tls] error: unexpected EOF");
+            } else {
+                ERR_error_string(ret, errbuf);
```
Comment (nokute78) [inline src/tls/openssl.c:446]: Ditto.

File: src/tls/openssl.c [hunk 4]
```diff
@@ -426,6 +437,15 @@ static int tls_net_handshake(struct flb_tls *tls, void *ptr_session)
         if (ret != SSL_ERROR_WANT_READ &&
             ret != SSL_ERROR_WANT_WRITE) {
             ret = SSL_get_error(session->ssl, ret);
+            // The SSL_ERROR_SYSCALL with errno value of 0
```
Comment (nokute78) [inline src/tls/openssl.c:440]: https://github.com/fluent/fluent-bit/blob/master/CONTRIBUTING.md#multiline-comment
Multiline comment should be
```
/*
 * This is my comment which is longer than 80 characters, so we must use the
 * multi-line type comments.
 */
```

Comment (nokute78) [inline src/tls/openssl.c:440]: https://github.com/fluent/fluent-bit/blob/master/CONTRIBUTING.md#multiline-comment
Multiline comment should be
```
/*
 * This is my comment which is longer than 80 characters, so we must use the
 * multi-line type comments.
 */
```

**Review by nokute78 — COMMENTED** [review]
I added some review comments.

- How about using thread-safe function `ERR_error_string_n` ?
- Multiline comment style

File: src/tls/openssl.c (hunk 1 above)
Comment (pwhelan) [reply src/tls/openssl.c]: I changed all uses of ERR_error_string to ERR_error_string_n.

File: src/tls/openssl.c (hunk 4 above)
Comment (pwhelan) [reply src/tls/openssl.c]: I reformated this comment to align to 80 columns.

**Comment by pwhelan** [comment]
@nokute78 or @edsiper any chance of getting this merged?

**Pull request merged**